Create archives in order, and upload and delete directly after creation

manuelwedler opened the pull request:
The archive job was creating the zip files on the NFS in GCP. To reduce load on the NFS, we want to keep the archives in memory of Cloud Run (same approach as in the parquet export). For this, it is needed to create, upload and delete the archives in order.

manuelwedler committed:
Create archives in order, and upload and delete directly after creation

kuzdogan commented on src/RepositoryChainsArchiver.ts:
```diff
@@ -51,105 +63,107 @@ export default class RepositoryChainsArchiver {
+            if (currentTarStream) {
+              currentTarStream.finalize();
+
+              // Upload file
+              const s3Key = `${backupName}/${archiveName}`;
+              await this.uploadFile(localPath, s3Key);
+              uploadedFiles.push({
+                path: s3Key,
+                sizeInBytes: fs.statSync(localPath).size,
+              });
```
Small comment, if the Job still does not finish under 24h, we can optimize here a bit. We can already start processing the next batch while it's uploading. In that sense it might be useful to put some logs here and in processing each 0xab entry to see how long it takes, and see the bottlenecks.

manuelwedler replied:
Yeah, might be worth to add some benchmarking. Let's see how it performs now. I also increased the Cloud Run CPU and memory a bit. I don't expect the upload to take too long, because the archives are in the size of around 12 MB.

kuzdogan commented on src/RepositoryChainsArchiver.ts:
```diff
@@ -51,105 +63,107 @@ export default class RepositoryChainsArchiver {
-      for (const archiveName of Object.keys(this.tarStreams)) {
-        this.tarStreams[archiveName].finalize();
       }
+      this.uploadManifest(backupName, uploadedFiles);
+      this.deleteOldBackups();
```
We need to await, no?

```suggestion
      await this.uploadManifest(backupName, uploadedFiles);
      await this.deleteOldBackups();
```

manuelwedler replied:
good catch :+1:

kuzdogan commented on src/RepositoryChainsArchiver.ts:
```diff
@@ -209,80 +219,33 @@ export default class RepositoryChainsArchiver {
-      // Clean up old backups
-      await this.deleteOldBackups(s3Client, s3Config.bucketName);
+    // Upload manifest files
+    await this.uploadFile(manifestPath, `${backupName}/manifest.json`);
+    await this.uploadFile(manifestPath, "manifest.json");
```
Why are there two uploads?

manuelwedler replied:
The bucket currently has this structure:
```
sourcify-repository-2024-12-17T00-00-21/
    -- manifest.json
    -- [other files]
sourcify-repository-2024-12-22T00-00-20/
    -- manifest.json
    -- [other files]
sourcify-repository-2024-12-23T00-00-26/
    -- manifest.json
    -- [other files]
manifest.json
```
So the manifest is duplicated, because we keep 3 backups. the manifest.json on the highest level of the bucket refers to the newest backup always. Let me know if you think anything of this should be changed.

kuzdogan replied:
Is the root manifest.json exactly the same as <latest-date>/manifest.json? Or is the top manifest a manifest of manifests? 

IMO we should redirect [repo-backup.sourcify.dev](https://repo-backup.sourcify.dev/) to the latest manifest, meaning root and latest should be duplicates. But we should document that the prev 2 backups are also available and can be seen with `ls`.

manuelwedler replied:
yes they are exactly the same

manuelwedler replied:
added a task to https://github.com/ethereum/sourcify/issues/1695 for documentation

kuzdogan replied:
A small thing then, if they are exactly the same, is to make the paths absolute in manifest.json:

```js
    {
      // "path": "sourcify-repository-2024-12-23T03-00-23/sourcify-repository-2024-12-23T03-00-23.part.gz.aa",
      "path": "/sourcify-repository-2024-12-23T03-00-23/sourcify-repository-2024-12-23T03-00-23.part.gz.aa",
      "sizeInBytes": 2097152000
    },
```

With that both manifests will lead to the same full URL instead of potentially interpreting this as relative paths.

https://repo-backup.sourcify.app/sourcify-repository-2024-12-23T03-00-23/sourcify-repository-2024-12-23T03-00-23.part.gz.aa

manuelwedler replied:
ah yes that should be fixed

kuzdogan commented on src/RepositoryChainsArchiver.ts:
```diff
@@ -51,105 +63,107 @@ export default class RepositoryChainsArchiver {
-    await s3Client.send(
+    await this.s3Client.send(
       new PutObjectCommand({
-        Bucket: bucketName,
+        Bucket: this.s3BucketName,
         Key: s3Key,
         Body: fileContent,
       })
     );
-    console.log(`Uploaded ${s3Key} to ${bucketName}`);
+    console.log(`Uploaded ${s3Key} to ${this.s3BucketName}`);
```
Another improvement might be creating a stream here instead of reading the whole file and uploading. But I assume this does not take too long? Logs again would help.

kuzdogan commented on src/RepositoryChainsArchiver.ts:
```diff
@@ -51,105 +63,107 @@ export default class RepositoryChainsArchiver {
+    const listResponse = await this.s3Client.send(
       new ListObjectsV2Command({
-        Bucket: bucketName,
+        Bucket: this.s3BucketName,
         Prefix: "sourcify-repository-",
         Delimiter: "/", // This will only list the root folders
```
Doesn't this potentiall delete the new backup? We are calling `deleteOldBackups` after uploading the new one. I couldn't see how we make the distinction between old and new backups in the bucket

manuelwedler replied:
Look further below. The function always keeps the latest three backups including the new one.

kuzdogan replied:
👍

kuzdogan reviewed (changes requested):
Some small comments and potential issues

manuelwedler commented:
The first trial run on staging went through succesfully without hitting the 24h limit. I wouldn't implement any performance improvement here now.

manuelwedler committed:
Await `uploadManifest` and `deleteOldBackups` functions

manuelwedler committed:
Make manifest paths absolute

Needed for the manifest files inside folders, otherwise the paths could be interpreted relatively

manuelwedler commented:
Made the small fixes. Do you think we can merge this? @kuzdogan

manuelwedler commented:
Had to change the s3Key again, because the additional `/` would add an additional folder to R2 and break the `deleteOldBackups` function

kuzdogan reviewed (approved):
LGTM

How long did it take to run the whole thing? I guess on staging HDD? @manuelwedler

manuelwedler commented:
It was 18 hours

OmarTawfik commented:
@manuelwedler Thank you so much for working on this!
 I wonder if the new format is available to try out anywhere already? or if not, is there an ETA for publishing this?

manuelwedler commented:
@OmarTawfik You can try our staging environment here: https://repo-backup.staging.sourcify.app/

For production, we still need to improve the performance of the script, since it is much more data. I hope production will be available next week.
